File: frontend/src/lib/api/safety-evals.ts
import { apiRequest } from "@/lib/api/request";

export type SafetyViolationType =
  | "injection"
  | "cross_tenant_leakage"
  | "private_source_exposure"
  | "unsupported_claims"
  | "malicious_document"
  | "unsafe_transform";

export type SafetyEvalSeverity = "critical" | "high" | "medium" | "low";

export type SafetyEvalCaseResponse = {
  case_id: string;
  suite_name: string;
  violation_type: SafetyViolationType;
  name: string;
  description: string | null;
  prompt_text: string;
  severity: SafetyEvalSeverity;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type SafetyEvalCaseListResponse = {
  items: SafetyEvalCaseResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type CreateSafetyEvalCaseRequest = {
  suite_name: string;
  violation_type: SafetyViolationType;
  name: string;
  prompt_text: string;
  severity?: SafetyEvalSeverity;
  description?: string | null;
  metadata?: Record<string, unknown>;
};

export type SafetyEvalRunResponse = {
  run_id: string;
  status: "queued" | "running" | "completed" | "failed";
  suite_name: string | null;
  pass_count: number | null;
  fail_count: number | null;
  total_count: number | null;
  pass_rate: number | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type SafetyEvalRunListResponse = {
  items: SafetyEvalRunResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type SafetyEvalResultResponse = {
  result_id: string;
  case_id: string;
  case_name: string;
  suite_name: string;
  violation_type: string;
  severity: string;
  passed: boolean;
  violation_detected: boolean;
  violation_type_detected: string | null;
  score: number | null;
  latency_ms: number | null;
  details: Record<string, unknown>;
  created_at: string;
};

export type SafetyEvalResultListResponse = {
  items: SafetyEvalResultResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type SafetyEvalRunDetailResponse = {
  run_id: string;
  status: "queued" | "running" | "completed" | "failed";
  suite_name: string | null;
  config: Record<string, unknown>;
  pass_count: number | null;
  fail_count: number | null;
  total_count: number | null;
  pass_rate: number | null;
  summary: Record<string, unknown>;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  results: SafetyEvalResultListResponse;
};

export type TriggerSafetyEvalRunRequest = {
  suite_name?: string | null;
  model_version?: string | null;
  retrieval_settings?: Record<string, unknown>;
  regression_threshold?: number | null;
};

export type TriggerSafetyEvalRunResponse = {
  run_id: string;
  status: string;
  message: string;
};

export type SafetyEvalReportResponse = {
  run_id: string;
  status: string;
  generated_at: string;
  suite_name: string | null;
  total_cases: number;
  pass_count: number;
  fail_count: number;
  pass_rate: number;
  baseline_pass_rate: number | null;
  regression_detected: boolean;
  regression_threshold: number | null;
  by_violation_type: Record<string, Record<string, unknown>>;
  by_severity: Record<string, Record<string, unknown>>;
  failed_cases: Array<Record<string, unknown>>;
  summary: Record<string, unknown>;
};

export async function createSafetyEvalCase(
  payload: CreateSafetyEvalCaseRequest,
): Promise<SafetyEvalCaseResponse> {
  return apiRequest<SafetyEvalCaseResponse>("/safety-evals/cases", {
    method: "POST",
    json: payload,
  });
}

export async function listSafetyEvalCases(
  params: {
    suite_name?: string;
    violation_type?: string;
    limit?: number;
    offset?: number;
  } = {},
): Promise<SafetyEvalCaseListResponse> {
  return apiRequest<SafetyEvalCaseListResponse>("/safety-evals/cases", {
    query: {
      suite_name: params.suite_name,
      violation_type: params.violation_type,
      limit: params.limit,
      offset: params.offset,
    },
  });
}

export async function triggerSafetyEvalRun(
  payload: TriggerSafetyEvalRunRequest,
): Promise<TriggerSafetyEvalRunResponse> {
  return apiRequest<TriggerSafetyEvalRunResponse>("/safety-evals/runs", {
    method: "POST",
    json: payload,
  });
}

export async function listSafetyEvalRuns(
  params: { suite_name?: string; limit?: number; offset?: number } = {},
): Promise<SafetyEvalRunListResponse> {
  return apiRequest<SafetyEvalRunListResponse>("/safety-evals/runs", {
    query: {
      suite_name: params.suite_name,
      limit: params.limit,
      offset: params.offset,
    },
  });
}

export async function getSafetyEvalRunDetail(
  runId: string,
  params: { limit?: number; offset?: number } = {},
): Promise<SafetyEvalRunDetailResponse> {
  return apiRequest<SafetyEvalRunDetailResponse>(
    `/safety-evals/runs/${encodeURIComponent(runId)}`,
    {
      query: {
        limit: params.limit,
        offset: params.offset,
      },
    },
  );
}

export async function getSafetyEvalReport(
  runId: string,
): Promise<SafetyEvalReportResponse> {
  return apiRequest<SafetyEvalReportResponse>(
    `/safety-evals/runs/${encodeURIComponent(runId)}/report`,
  );
}

File: backend/app/domains/safety_evals/schemas/safety_evals.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator


class CreateSafetyEvalCaseRequest(BaseModel):
    suite_name: str = Field(..., min_length=1, max_length=255)
    violation_type: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=255)
    prompt_text: str = Field(..., min_length=1)
    severity: str = Field(default="high", max_length=32)
    description: str | None = Field(default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("violation_type")
    @classmethod
    def validate_violation_type(cls, v: str) -> str:
        allowed = {
            "injection",
            "cross_tenant_leakage",
            "private_source_exposure",
            "unsupported_claims",
            "malicious_document",
            "unsafe_transform",
        }
        if v not in allowed:
            raise ValueError(f"violation_type must be one of {sorted(allowed)}")
        return v

    @field_validator("severity")
    @classmethod
    def validate_severity(cls, v: str) -> str:
        allowed = {"critical", "high", "medium", "low"}
        if v not in allowed:
            raise ValueError(f"severity must be one of {sorted(allowed)}")
        return v


class SafetyEvalCaseResponse(BaseModel):
    case_id: str
    suite_name: str
    violation_type: str
    name: str
    description: str | None
    prompt_text: str
    severity: str
    metadata: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class SafetyEvalCaseListResponse(BaseModel):
    items: list[SafetyEvalCaseResponse]
    total: int
    limit: int
    offset: int


class TriggerSafetyEvalRunRequest(BaseModel):
    suite_name: str | None = Field(default=None, max_length=255)
    model_version: str | None = Field(default=None, max_length=128)
    retrieval_settings: dict[str, Any] = Field(default_factory=dict)
    regression_threshold: float | None = Field(default=None, ge=0.0, le=1.0)


class SafetyEvalRunResponse(BaseModel):
    run_id: str
    status: str
    suite_name: str | None
    pass_count: int | None
    fail_count: int | None
    total_count: int | None
    pass_rate: float | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class SafetyEvalRunListResponse(BaseModel):
    items: list[SafetyEvalRunResponse]
    total: int
    limit: int
    offset: int


class SafetyEvalResultResponse(BaseModel):
    result_id: str
    case_id: str
    case_name: str
    suite_name: str
    violation_type: str
    severity: str
    passed: bool
    violation_detected: bool
    violation_type_detected: str | None
    score: float | None
    latency_ms: int | None
    details: dict[str, Any]
    created_at: datetime


class SafetyEvalRunDetailResponse(BaseModel):
    run_id: str
    status: str
    suite_name: str | None
    config: dict[str, Any]
    pass_count: int | None
    fail_count: int | None
    total_count: int | None
    pass_rate: float | None
    summary: dict[str, Any]
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    results: SafetyEvalResultListResponse


class SafetyEvalResultListResponse(BaseModel):
    items: list[SafetyEvalResultResponse]
    total: int
    limit: int
    offset: int


class TriggerSafetyEvalRunResponse(BaseModel):
    run_id: str
    status: str
    message: str


class SafetyEvalReportResponse(BaseModel):
    run_id: str
    status: str
    generated_at: datetime
    suite_name: str | None
    total_cases: int
    pass_count: int
    fail_count: int
    pass_rate: float
    baseline_pass_rate: float | None
    regression_detected: bool
    regression_threshold: float | None
    by_violation_type: dict[str, dict[str, Any]]
    by_severity: dict[str, dict[str, Any]]
    failed_cases: list[dict[str, Any]]
    summary: dict[str, Any]
